rosbag2_performance/rosbag2_performance_benchmarking/src/benchmark_publishers.cpp

```diff
@@ -109,20 +117,34 @@ class BenchmarkPublishers : public rclcpp::Node
     producer->period = std::chrono::milliseconds(
       producer_config.frequency ? 1000 / producer_config.frequency : 1);
 
+    thread_pool_.queue(
+      [this, initial_time, producer] {
+        producer_job(initial_time, producer);
+      });
+
     return producer;
   }
 
-  void producer_job(std::shared_ptr<BenchmarkProducer> producer)
+  void producer_job(
+    std::chrono::time_point<std::chrono::high_resolution_clock> when,
+    std::shared_ptr<BenchmarkProducer> producer)
   {
-    for (auto i = 0u; i < producer->max_messages; ++i) {
-      if (!rclcpp::ok()) {
-        break;
-      }
-      std::this_thread::sleep_for(producer->period);
-      producer->produce();
+    std::this_thread::sleep_until(when);
+    producer->produce();
+
+    if (producer->produced_messages < producer->max_messages) {
+      thread_pool_.queue(
+        [this, next = std::chrono::high_resolution_clock::now() + producer->period,
+        producer] {
+          producer_job(next, producer);
+        });
```

Comment:
```suggestion
        [this, next_timestamp = when + producer->period, producer] {
          producer_job(next_timestamp, producer);
        });
```
This is interesting finding! 
Producer period means interval between messages. Call for the `producer->produce();` could take some substantial amount of time which we shouldn't count when calculating next timestamp.

Reply:
You are right, we can do it as you suggests, use an initial time and increase period each time. One relevant change is that now if a job last longer than one period the next time will be in the past and there will be no wait. This is probably better than the previous behavior but I wonder if we should warn the user this is happening.

Reply:
I think this is better approach since it is more close to the real use case.
In reality most system will behave this way since also have some internal buffering on upper layers.
If we will keep it as is we can't guaranty that if we define message rate 100 frames per second that it will be published exact 100 frames in one second. Most probably it will be published 99 or less.